merlin/utils.py

```diff
@@ -497,6 +498,31 @@ def contains_shell_ref(string):
     return False
 
 
+def dict_deep_merge(dict_a, dict_b, path=None):
+    """
+    This function recursively merges dict_b into dict_a. The built-in
+    merge of dictionaries in python (dict(dict_a) | dict(dict_b)) does not do a
+    deep merge so this function is necessary. Credit to this stack
+    overflow post: https://stackoverflow.com/a/7205107.
+
+    :param `dict_a`: A dict that we'll merge dict_b into
+    :param `dict_b`: A dict that we want to merge into dict_a
+    :param `path`: The path down the dictionary tree that we're currently at
+    """
+    if path is None:
+        path = []
+    for key in dict_b:
+        if key in dict_a:
+            if isinstance(dict_a[key], dict) and isinstance(dict_b[key], dict):
+                dict_deep_merge(dict_a[key], dict_b[key], path + [str(key)])
+            elif dict_a[key] == dict_b[key]:
+                pass  # same leaf value
+            else:
+                raise Exception(f"Conflict at {'.'.join(path + [str(key)])}")
```

Comment:
A more custom exception type should be used here I think.  Raising base exception type makes it hard to catch and handle just this one without also catching any other unrelated exception thrown in this scope.  Could be as simple as:

```
class DeepMergeException(Exception):
    pass
```
Or you can extend it to take more args and do more custom messaging (may need to implement `__reduce__` method too to ensure it can be pickled if you start adding more args).